                                                  <<<  ***Data understanding and cleaning.*** >>>

---


In [4]:
#import the required libraries for data 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [17]:
#load the data set
df = pd.read_csv('/kaggle/input/ibm-hr-analytics-attrition-dataset/WA_Fn-UseC_-HR-Employee-Attrition.csv')
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [7]:
#The size of the data set
print('The total length of the dataset is: ',df.shape[0])
print('The total features are: ',df.shape[1])

The total length of the dataset is:  1470
The total features are:  35


In [8]:
#colums in the dataset
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [11]:
for i,j in enumerate(df.columns):
    print(i+1,': ',j )

1 :  Age
2 :  Attrition
3 :  BusinessTravel
4 :  DailyRate
5 :  Department
6 :  DistanceFromHome
7 :  Education
8 :  EducationField
9 :  EmployeeCount
10 :  EmployeeNumber
11 :  EnvironmentSatisfaction
12 :  Gender
13 :  HourlyRate
14 :  JobInvolvement
15 :  JobLevel
16 :  JobRole
17 :  JobSatisfaction
18 :  MaritalStatus
19 :  MonthlyIncome
20 :  MonthlyRate
21 :  NumCompaniesWorked
22 :  Over18
23 :  OverTime
24 :  PercentSalaryHike
25 :  PerformanceRating
26 :  RelationshipSatisfaction
27 :  StandardHours
28 :  StockOptionLevel
29 :  TotalWorkingYears
30 :  TrainingTimesLastYear
31 :  WorkLifeBalance
32 :  YearsAtCompany
33 :  YearsInCurrentRole
34 :  YearsSinceLastPromotion
35 :  YearsWithCurrManager


**Observation**: There are 35 columns

In [12]:
#let us make the attrition as last feature for our convenience
df = df[['Age', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager', 'Attrition']]

In [13]:
#checking for missing values
df.isnull().sum()

Age                         0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
YearsWithC

**Observation**:  This data set doesn't contain null values.

In [14]:
#Let us observe the brief information of the data set
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   BusinessTravel            1470 non-null   object
 2   DailyRate                 1470 non-null   int64 
 3   Department                1470 non-null   object
 4   DistanceFromHome          1470 non-null   int64 
 5   Education                 1470 non-null   int64 
 6   EducationField            1470 non-null   object
 7   EmployeeCount             1470 non-null   int64 
 8   EmployeeNumber            1470 non-null   int64 
 9   EnvironmentSatisfaction   1470 non-null   int64 
 10  Gender                    1470 non-null   object
 11  HourlyRate                1470 non-null   int64 
 12  JobInvolvement            1470 non-null   int64 
 13  JobLevel                  1470 non-null   int64 
 14  JobRole                 

***Observation:*** There are 26 numerical features and 9 categorical features.

But as we observe here some columns are miss classified as numerical while they are actually categorical. Lets explore them in next step.

In [6]:
#Now let us get some information about The total unique values in a feature
for i,j in enumerate(df.columns):
    print(f'for column {i+1} {j}: ')
    print(f'There are {df[j].nunique()} unique values')
    print(f'Data type of {j}: {df[j].dtype}')
    if df[j].nunique() > 30:
        print('Many unique values')
    else:    
        print('They are: ',df[j].unique())
    print('==========================================================================')
    print()

for column 1 Age: 
There are 43 unique values
Data type of Age: int64
Many unique values

for column 2 Attrition: 
There are 2 unique values
Data type of Attrition: object
They are:  ['Yes' 'No']

for column 3 BusinessTravel: 
There are 3 unique values
Data type of BusinessTravel: object
They are:  ['Travel_Rarely' 'Travel_Frequently' 'Non-Travel']

for column 4 DailyRate: 
There are 886 unique values
Data type of DailyRate: int64
Many unique values

for column 5 Department: 
There are 3 unique values
Data type of Department: object
They are:  ['Sales' 'Research & Development' 'Human Resources']

for column 6 DistanceFromHome: 
There are 29 unique values
Data type of DistanceFromHome: int64
They are:  [ 1  8  2  3 24 23 27 16 15 26 19 21  5 11  9  7  6 10  4 25 12 18 29 22
 14 20 28 17 13]

for column 7 Education: 
There are 5 unique values
Data type of Education: int64
They are:  [2 1 4 3 5]

for column 8 EducationField: 
There are 6 unique values
Data type of EducationField: object
T

Form the above observation, The data should be subjected to data cleaning.

#                                                       Data Cleaning

By this we can determine some of the columns that are not important for analysing and modelling.

They are:

1. EmployeeCount: This feature has only one unique value which is not important.
2. EmployeeNumber: This feature is just a unique number of the employees.
3. Over18: since all the employees are over18 there is no need of this feature.
4. StandardHours: for evey employee the standard hours are 80.

These columns are to be removed.


There are some columns initially numerical but they are actually categorical, They should be replaced by a suitable names.

They are:

1. Education
2. EnvironmentSatisfaction
3. JobInvolvement
4. JobLevel
5. JobSatisfaction
6. PerformanceRating
7. RelationshipSatisfaction
8. WorkLifeBalance
9. StockOptionLevel

For the StockOptionLevel describes High-level employees of a company, such as executives or managers who are typically given the option to purchase shares of the company at a set price, which is usually lower than the market price. So it basically gives the level of StockOption. It may come under categorical column which is of Level 0, Level 1, Level 2, Level 3.

In [18]:
df = df.drop(columns = ['EmployeeCount','EmployeeNumber','Over18','StandardHours'], axis = 1)

In [19]:
df['Education'] = df['Education'].apply(lambda x : 'Schooling' if x == 1 else ( 'Intermediate' if x == 2 else ( 'Bachelors' if x == 3 else ('Masters' if x == 4 else 'PhD'))))

In [20]:
df['EnvironmentSatisfaction'] = df['EnvironmentSatisfaction'].apply(lambda x : 'Low' if x == 1 else ( 'Medium' if x == 2 else ( 'High' if x == 3 else 'Very high')))

In [21]:
df['JobInvolvement'] = df['JobInvolvement'].apply(lambda x : 'Low' if x == 1 else ( 'Medium' if x == 2 else ( 'High' if x == 3 else 'Very high')))

In [22]:
df['JobLevel'] = df['JobLevel'].apply(lambda x : 'Fresher' if x == 1 else ( 'Junior' if x == 2 else ( 'Senior' if x == 3 else ('Team Lead' if x == 4 else 'Manager'))))

In [23]:
df['JobSatisfaction'] = df['JobSatisfaction'].apply(lambda x : 'Low' if x == 1 else ( 'Medium' if x == 2 else ( 'High' if x == 3 else 'Very high')))

In [24]:
df['PerformanceRating'] = df['PerformanceRating'].apply(lambda x : 'Poor' if x == 1 else ( 'Good' if x == 2 else ( 'Excellent' if x == 3 else 'Outstanding')))

In [25]:
df['RelationshipSatisfaction'] = df['RelationshipSatisfaction'].apply(lambda x : 'Poor' if x == 1 else ( 'Good' if x == 2 else ( 'Excellent' if x == 3 else 'Outstanding')))

In [26]:
df['WorkLifeBalance'] = df['WorkLifeBalance'].apply(lambda x : 'Poor' if x == 1 else ( 'Good' if x == 2 else ( 'Better' if x == 3 else 'Best')))

In [27]:
df['StockOptionLevel'] = df['StockOptionLevel'].apply(lambda x : 'Level 0' if x == 0 else ( 'Level 1' if x == 1 else ( 'Level 2' if x == 3 else 'Level 3')))

In [28]:
#The size of the data set
print('The total length of the dataset is: ',df.shape[0])
print('The total features are: ',df.shape[1])

The total length of the dataset is:  1470
The total features are:  31


***Observation:***

Now the shape of this new data set is:   
     1.1470 rows with 31 features.

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 31 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   object
 7   EducationField            1470 non-null   object
 8   EnvironmentSatisfaction   1470 non-null   object
 9   Gender                    1470 non-null   object
 10  HourlyRate                1470 non-null   int64 
 11  JobInvolvement            1470 non-null   object
 12  JobLevel                  1470 non-null   object
 13  JobRole                   1470 non-null   object
 14  JobSatisfaction         

***Observation:*** Now there are 14 numerical columns and 17 categorical columns.

In [31]:
#statistal description of numerical columns
df.describe().T

count          mean          std     min     25%  \
Age                      1470.0     36.923810     9.135373    18.0    30.0   
DailyRate                1470.0    802.485714   403.509100   102.0   465.0   
DistanceFromHome         1470.0      9.192517     8.106864     1.0     2.0   
HourlyRate               1470.0     65.891156    20.329428    30.0    48.0   
MonthlyIncome            1470.0   6502.931293  4707.956783  1009.0  2911.0   
MonthlyRate              1470.0  14313.103401  7117.786044  2094.0  8047.0   
NumCompaniesWorked       1470.0      2.693197     2.498009     0.0     1.0   
PercentSalaryHike        1470.0     15.209524     3.659938    11.0    12.0   
TotalWorkingYears        1470.0     11.279592     7.780782     0.0     6.0   
TrainingTimesLastYear    1470.0      2.799320     1.289271     0.0     2.0   
YearsAtCompany           1470.0      7.008163     6.126525     0.0     3.0   
YearsInCurrentRole       1470.0      4.229252     3.623137     0.0     2.0   
YearsSinceLastPromotion  1470.0      2.187755     3.222430     0.0     0.0   
YearsWithCurrManager     1470.0      4.123129     3.568136     0.0     2.0   

                             50%       75%      max  
Age                         36.0     43.00     60.0  
DailyRate                  802.0   1157.00   1499.0  
DistanceFromHome             7.0     14.00     29.0  
HourlyRate                  66.0     83.75    100.0  
MonthlyIncome             4919.0   8379.00  19999.0  
MonthlyRate              14235.5  20461.50  26999.0  
NumCompaniesWorked           2.0      4.00      9.0  
PercentSalaryHike           14.0     18.00     25.0  
TotalWorkingYears           10.0     15.00     40.0  
TrainingTimesLastYear        3.0      3.00      6.0  
YearsAtCompany               5.0      9.00     40.0  
YearsInCurrentRole           3.0      7.00     18.0  
YearsSinceLastPromotion      1.0      3.00     15.0  
YearsWithCurrManager         3.0      7.00     17.0

***Observation:*** 

The above is the description of the mean, standard deviation, quartiles, minimum and maximum values for the numerical data.
From this many statistical insights are gained.

- from age:
    - 1. All the employees are majors.
      2. age is well distributed and may not contains any outliers.

- from daily rate:
    - 1. it is also well distributed.
      2. highest daily rate is 1499 and the lowest is 102 while 802.4 .

- from DistanceFromHome:
    - 1. Most of the employees are near to their work place.
      2. 75% of the employee homes are below 15 Kms.
      3. It can be right skewed.

- from HourlyRate:
    - 1. The data is Normally distributed.
      2. The maximum hourly rate is 100 and minmimum is 30 while the mean is 65.8 .

- from MonthlyIncome:
    - 1. from the above observation we can clearly determine that there are some outliers beacause of the huge difference between 3rd quartile and maximum value.
      2. The data can be right skewed.
      3. The highest salary is 19999 while the 50% of the employees falls under the range of 1009 to 4919 monthly income.

- from MonthlyRate:
    - 1. The maximum monthly rate is 26999, The minimum is 2094 while mean is 14313.
      2. The data is normally distributed.

- from NumCompaniesWorked:
    - 1. 50% of the employees are worked less than 3 companies.
      2. The data can be right skewed.

- from PercentSalaryHike:
    - 1. The minimum 11% of salary hike is given by the company the maximum is 25.
      2. 50% of the employees got only 14% hike.

- from TotalWorkingYears:
    - 1. 50% of the employees are below 11 years experience while the maximum is of 40 years experience.
      2. As the 3rd quartile is 15 and maximum is of 40, it may contain some outliers.

- from TrainingTimesLastYear:
    - 1. 75% of the employees are trained below 4 years, but the the maximum is of 6 years.
      2. As there is huge difference between 3rd quartile and maximum value it may contain some outliers.

- from YearsAtCompany	:
    - 1. 50% of the employees are below 6 years in the present company where the maximum is of 40 years.
      2. may contain outliers and the data is right skewed.
    
- form YearsInCurrentRole:
    - 1. 50% of the employees are only of 3 years on their current role.
      2. The maximum is of 18 years.
      3. may contain outliers and the data is right skewed.

- from YearsSinceLastPromotion:
    - 1. 75% of the employees are promoted at 3years but the maximum goes up to 15.
      2. may contain outliers and the data is right skewed.

- from YearsWithCurrManager:
    - 1. for 50% of the employees manager have changed for 3 years while the maximum is of 17 years.
      2. The data is right skewed and may contain some outliers.

- The questions to be raised from this observations are:
    - 1. What is the attrition rate of the company?
      2. In which category of business travel the attrition rate is higher?
      3. In which department highest attrition took place?
      4. How does the education impacts attrition?
      5. Do the employess with less hike are likely to increase the attrition rate?
      6. In which department the bachelor degrees are more?
      7. How does the previous worked number of companies impact attrition?
      8. Which age category does the attrition rate is more?
      9. How does the maritial status and gender impacts attrition?
      10. Does the attrition rate is more for less hike employees?
      11. In which education field the attrition is more?
      12. Is environment satisfaction impacts attrition rate?
      13. Do the employees with less salary are likely to increase the attrition rate?
      14. Which gender is likely to stay in the company?
      15. Does the marriage status impacts attrition rate?
      16. Which job level has a great job involvement?
      17. Does the employees with long distance to their home are likely to increase attrition rate?
      18. Does the year spent at same company increase attrition rate?
      19. Does the low job involvement had a chance to increase attrition rate?
      20. Which level of job has higher attrition rate?
      21. Which job role has higher attrition?
      22. Does the low job satisfaction leads to increase in attrition rate?
      23. Which job level employee are likely to do over time?
      24. Does the over time increases in attrition rate?
      25. Which category of performance rating has higher attrition rate?
      26. Does the poor relationship satisfaction had more attrition rate?
      27. Does an employee likely to leave a company with poor work life balance?
      28. What are the features that have a good correlation?

These are some questions raised by observing the data. There can be many more question to be raised according to the business need and requirement.

In [33]:
#determine numeric and categorical columns
numerical_cols = df.select_dtypes(include = 'int')
categorical_cols = df.select_dtypes(include = 'object')

In [34]:
#export the prepared data set
df.to_csv('HR-Employee-Attrition-cleaned-data.csv', index=False)